src/image/pixels.js

```diff
@@ -752,34 +753,12 @@ function pixels(p5, fn){
 
     // when this is P2D renderer, create/use hidden webgl renderer
     else {
-      const filterGraphicsLayer = this.getFilterGraphicsLayer();
-      // copy p2d canvas contents to secondary webgl renderer
-      // dest
-      filterGraphicsLayer.copy(
-        // src
-        this._renderer,
-        // src coods
-        0, 0, this.width, this.height,
-        // dest coords
-        -this.width/2, -this.height/2, this.width, this.height
-      );
-      //clearing the main canvas
-      this._renderer.clear();
-
-      this._renderer.resetMatrix();
-      // filter it with shaders
-      filterGraphicsLayer.filter(...args);
-
-      // copy secondary webgl renderer back to original p2d canvas
-      this.copy(
-        // src
-        filterGraphicsLayer._renderer,
-        // src coods
-        0, 0, this.width, this.height,
-        // dest coords
-        0, 0, this.width, this.height
-      );
-      filterGraphicsLayer.clear(); // prevent feedback effects on p2d canvas
+    
+     if (!this.filterRenderer) { 
+        this.filterRenderer = new FilterRenderer2D(this, operation, value);     
+      }   
+             
+      this.filterRenderer.applyFilter();
```

Comment:
If a different filter is rendered (e.g. a blur then a threshold) I think we'll need to update the addressee used here. Might need to add an update method, and maybe also cache the shaders by their type instead of just having one cached

Reply:
And I guess if a custom filter shader is passed in, we don't need to cache that, since the user holds on to the shader instance themself in that case